File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/CoinSelectionAdapter.java
package com.brentpanther.bitcoinwidget;

import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.Collections;
import java.util.EnumSet;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class CoinSelectionAdapter extends RecyclerView.Adapter<CoinSelectionAdapter.ViewHolder> {

    private CoinSelectionView.CoinSelectedListener listener;
    private final ArrayList<Coin> coins;

    static class ViewHolder extends RecyclerView.ViewHolder {

        ViewHolder(CoinSelectionView v) {
            super(v);
        }
    }

    CoinSelectionAdapter(CoinSelectionView.CoinSelectedListener listener) {
        this.listener = listener;
        coins = new ArrayList<>(EnumSet.allOf(Coin.class));
        Collections.sort(coins, (o1, o2) -> o1.getName().compareTo(o2.getName()));
    }

    @NonNull
    @Override
    public CoinSelectionAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(new CoinSelectionView(parent.getContext()));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        ((CoinSelectionView)holder.itemView).setCoin(coins.get(position), listener);
    }

    @Override
    public int getItemCount() {
        return Coin.values().length;
    }
}

File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/CoinSelectionAdapter.kt
package com.brentpanther.bitcoinwidget

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import java.util.*

internal class CoinSelectionAdapter internal constructor(private val listener: ((coin: Coin) -> kotlin.Unit)?) : RecyclerView.Adapter<CoinSelectionAdapter.ViewHolder>() {

    private val coins: MutableList<Coin> = Coin.values().toMutableList()

    internal class ViewHolder(v: CoinSelectionView) : RecyclerView.ViewHolder(v)

    init {
        coins.sortWith(Comparator { o1, o2 -> o1.coinName.compareTo(o2.coinName) })
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(CoinSelectionView(parent.context))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        (holder.itemView as CoinSelectionView).setCoin(coins[position], listener)
    }

    override fun getItemCount(): Int {
        return Coin.values().size
    }
}